jdk/src/share/classes/java/security/SecureRandom.java

```diff
@@ -202,16 +202,8 @@ private void getDefaultPRNG(boolean setSeed, byte[] seed) {
         String prng;
 
         // If in FIPS mode, use the SecureRandom from the FIPS provider.
-        if (FIPSConfigurator.enableFIPS()) {
-            Provider p = Security.getProvider("SunPKCS11-NSS-FIPS");
-            prng = "PKCS11";
-            if (p == null) {
-                throw new RuntimeException("could not find SunPKCS11-NSS-FIPS provider for FIPS mode");
-            }
-            Service prngService = p.getService("SecureRandom", prng);
-            if (prngService == null) {
-                throw new RuntimeException("could not find SecureRandom implementation from SunPKCS11-NSS-FIPS");
-            }
+        if (RestrictedSecurity.isEnabled()) {
+            prng = RestrictedSecurity.getRandomAlgorithm();
```

Comment:
In JDK11 and above version, in the getDefaultPRNG() method, it uses the “provider.getService("SecureRandom", prngAlgorithm);” method to return the SecureRandom service. So, we added a RestrictedSecurity.getRandomProvider() method to return the allowed SecureRandom provider in the Java Restricted Security Mode.

But in JDK8, in the getDefaultPRNG() method, it uses the “SecureRandom.getInstance(prng)” method. And this method will traverse the list of registered security providers to return the SecureRandom service. So, in JDK8, no need to return the allowed SecureRandom provider by the RestrictedSecurity.getRandomProvider() method.

Reply:
When get the SecureRandom instance, it will traverse the list of registered security providers and return the first provider which has the SecureRandom algorithm “PKCS11”. And, only the provider SunPKCS11 has the SecureRandom algorithm “PKCS11”. So, the correct provider SunPKCS11 will be returned.

For the RestrictedSecurity.getRandomProvider() method, since we don’t need it in OpenJDK8, so I removed it from the RestrictedSecurity.java.